tmt/utils.py

```diff
@@ -404,6 +404,218 @@ class CommandOutput(NamedTuple):
+            start_new_session=True,
+            stdin=subprocess.DEVNULL, stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT if join else subprocess.PIPE,
+            executable=executable)
+    except FileNotFoundError as error:
+        raise RunError(
+            f"File '{error.filename}' not found.",
+            command,
+            127,
+            caller=caller)
+
+    stdout_thread = StreamLogger(
+        process.stdout,
+        log_header='out',
+        logger=log,
+        click_context=click.get_current_context(silent=True))
+    stderr_thread = stdout_thread
+    if not join:
+        stderr_thread = StreamLogger(
+            process.stderr,
+            log_header='err',
+            logger=log,
+            click_context=click.get_current_context(silent=True))
+    stdout_thread.start()
+    if not join:
+        stderr_thread.start()
+
+    # A bit of logging helpers for debugging duration behavior
+    start_timestamp = time.monotonic()
+
+    def _event_timestamp() -> str:
+        return f'{time.monotonic() - start_timestamp:.4}'
+
+    def log_event(msg: str) -> None:
+        logger.debug('Command event', f'{_event_timestamp()} {msg}', level=4)
+
+    try:
+        process.wait(timeout=timeout)
+    except subprocess.TimeoutExpired:
+        log_event(f'duration "{timeout}" exceeded')
+
+        os.killpg(os.getpgid(process.pid), signal.SIGKILL)
+        log_event('sent SIGKILL signal')
+
+        process.wait()
+        log_event('kill confirmed')
+
+        process.returncode = PROCESS_TIMEOUT
+
+    log_event('waiting for stream readers')
+
+    stdout_thread.join()
+    log_event('stdout reader done')
+
+    if not join:
+        stderr_thread.join()
+        log_event('stderr reader done')
+
+    # Handle the exit code, return output
+    if process.returncode != 0:
+        raise RunError(
+            message=f"Command returned '{process.returncode}'.",
+            command=command,
+            returncode=process.returncode,
+            stdout=stdout_thread.get_output(),
+            stderr=stderr_thread.get_output(),
+            caller=caller)
+    if join:
+        return CommandOutput(
+            stdout_thread.get_output(), None)
+    else:
+        return CommandOutput(
+            stdout_thread.get_output(), stderr_thread.get_output())
+
+
+def run_command(
+        *,
+        command: Command,
+        friendly_command: Optional[str] = None,
+        silent: bool = False,
+        message: Optional[str] = None,
+        cwd: Optional[Path] = None,
+        dryrun: bool = False,
+        suppress_dryrun: bool = False,
+        shell: bool = False,
+        env: Optional[EnvironmentType] = None,
+        interactive: bool = False,
+        join: bool = False,
+        log: Optional[BaseLoggerFnType] = None,
+        timeout: Optional[int] = None,
+        caller: Optional['Common'] = None,
+        logger: tmt.log.Logger) -> CommandOutput:
```

Comment:
Is the arg `logger` a must? If yes, what's the default logger we can use if we invoke:
```python
result = run_command(command=<MY COMMAND>, cwd=<MY CWD>, logger=<???>)
```
?

Reply:
(Sorry for the rather lengthy response, I'm fully aware this logger stuff is a serious change from how tmt did its logging, and it's probably highly uncomfortable and uncommon for the tmt developers - I take every opportunity to illustrate issues and and that I'm always willing to help to ease us all in, and that I'm not trying to complicate your patches just for fun :)

There's no default logger per se, no `tmt.log.DEFAULT_LOGGER` to use. The "run a command" is way too low in the food chain to be aware of the bigger picture, and with the hierarchy of various objects (run -> plans -> steps -> ...), it may be impossible to guess correctly. This means, callers of `run_command()` must provide the logger, hence missing the default value. `run()` method in `Common` uses `self._logger`, although some step plugins will use a different logger.

The `logger` needs to "infect" more functions. `default_branch()` gets a new `logger` parameter (it should be the last in the list), and all its call sites would need to pass a logger in. There are three such places, two do have a logger at hand, and the third is `tmt.utils.fmf_id` which suffers the same defect, it needs to log stuff but can be called from different levels of the tree above. Luckily, both call sites of `fmf_id` are in `tmt.base.Core`, which can hand over `self._logger` - it's not perfect, but it's much better than any default logger we could think of (`Core` is a base class of tests, plans, and stories, each has a different position in the tree, any default logger would be wrong for two of these three at best...).

I know it sounds needlessly complicated, and `logger` is viewed as a pesky addition, and I'm sorry about that, but I'm afraid the days of the simple single-threaded workflow are gone. Let me illustrate: https://github.com/teemtee/tmt/discussions/1702#discussioncomment-5233531 is an example of a multi-host run, with several tests running on different guests at the same time. To distinguish the output of different guests, custom `Logger` instances are created for each "thread", and it adds a label at the beginning of each line. But, then there are methods and functions that use some kind of "default" logger, usually `self.info()` or similar, instead of a logger provided by whoever calls them, and these lack the extra labels. `test: Dummy test C` - from which guest is this line, is it a client-1, client-2 or server? And that's just the visible output, then there's the logfile in which all levels of debug and verbose lines end up, and there we could find the full log of a `self.run(...)` call with no context attached - which guest the call belongs to?

`default_branch()` and `fmf_id` seem to be used exclusively from the `discover` step, which is not parallelized - **yet**. There's no technical reason against evaluating `discover` steps in parallel, so, maybe one day, who knows? This is why I'm pushing us collectively into building "purer" functions, to rely more on their arguments rather than mutable shared state - to build the habit, to let this concept put down its roots in our minds, that mutable shared state is often a complication: reading `self.data` is fine, but emitting output via `self._logger` might lead to unexpected results as other parts of tmt might do the same, and break our toys.

Reply:
looks resolved.